### PR TITLE
Enable C++ compiling and running in windows

### PR DESCRIPTION
**Pull request by TheBrenny**
Pretty much just copied and pasted. I edited this on GH, so I can't be 100% sure on if it works or not.

But with Microsoft's Insider Preview 14399, they have created a Linux Subsystem for Windows (or LSW). With this, installed, bash is automatically installed, and therefore is available for use from the command line. Better yet, it's available in the exact same way as you would use it on Linux because (as the promo video says) it's "bash running on Linux running on Windows. No Virtual Machine in sight."

Hopefully this will start a new age of not absolutely requiring Visual Studio when developing on Windows?

### CODE REVIEW

File: lib/grammars.coffee [hunk 1]
```diff
@@ -68,6 +68,10 @@ module.exports =
       "File Based":
         command: "bash"
         args: (context) -> ["-c", "g++ -Wall -include stdio.h -include iostream '" + context.filepath + "' -o /tmp/cpp.out && /tmp/cpp.out"]
+    else if GrammarUtils.OperatingSystem.isWindows() and GrammarUtils.OperatingSystem.release().indexOf("14399") >= 0
+      "File Based":
+        command: "bash"
```
Comment (nixel2007) [inline lib/grammars.coffee]: bash?

**Comment by idleberg** [comment]
@nixel2007 That would be [Bash on Ubuntu on Windows](https://msdn.microsoft.com/en-us/commandline/wsl/about) , a native bash for Windows 10. It comes with `apt-get` and let's you install `gcc` among other things.

@TheBrenny As far as I know, you need to translate the Windows path to a POSIX path, how would you do that? The equivalent of `C:\Users\Desktop\hello_world.cpp` is `/mnt/c/Users/Desktop/hello_world.cpp`, but I'm unaware of programatic ways to translate the path (e.g. like `cygpath` does on Cygwin)

**Comment by TheBrenny** [comment]
@idleberg With [my answer to your question](http://stackoverflow.com/a/38160347/2238442), it's possible that we can `require("path")` and use my one-liner method? I'll commit to my master branch and we'll go from there.

**Comment by TheBrenny** [comment]
I don't understand why the builds are failing? [The Travis CI logs](https://travis-ci.org/rgbkrk/atom-script/jobs/141857755#L277) show that it's because `GrammarUtils.OperatingSystem.release()` doesn't have a function attached to it called `slice` - which I don't understand why not? It *should*, shouldn't it?

**Comment by nixel2007** [comment]
Why we can't use simple `cmd` for this?

**Comment by nixel2007** [comment]
Current variant can run only on newest version of win. But there is a lot of win8 or win 7 users

**Comment by TheBrenny** [comment]
AFAIK, There is no "built-in" way to build a C++ file on windows. Microsoft
has provided WSL, which allows users to use Bash on windows, while also
utilising g++ for compiling.

**Comment by idleberg** [comment]
(Just to give some insight anybody unfamiliar with WSL)

> AFAIK, There is no "built-in" way to build a C++ file on windows.

That's True, but neither is gcc a "build-in" way. Bash on Ubuntu on Windows itself isn't installed by default, it requires a checkbox to be activated, much in the fashion Telnet is currently installed. Once installed, you still have to run `apt-get install gcc`.

**Comment by nixel2007** [comment]
Please, correct me if i'm wrong, but you can use a g++ via MinGW package?
The linux g++ (from apt) will not work trough casual cmd? you need to run it from built-in bash?

**Comment by nixel2007** [comment]
> Personally, I would prefer choice over which compiler to use, but I don't think it's possible in script, right? The build package, for instance let's you install build providers, so you can for instance choose whether to use sass to sassc to compile Sass/SCSS.

you can change the command in script options view (and save it as profile, if you prefer)

**Comment by nixel2007** [comment]
@TheBrenny may i ask you to fix the tests?
there are some "windows can't into c++" tests here.

**Comment by nixel2007** [comment]
Shipped in 3.8.0 (and 3.8.1) :ship:

**Comment by nixel2007** [comment]
> I don't understand why the builds are failing? The Travis CI logs show that it's because GrammarUtils.OperatingSystem.release() doesn't have a function attached to it called slice - which I don't understand why not? It should, shouldn't it?

Fixed in master

**Comment by 8fdafs2** [comment]
When using **script** to _compile and run_ (or just _compile_?) a cpp file on windows 10 pro 64bit I get the following output:
```
g++.exe: error: C:/Program Files/Git/mnt/d/somefolder/some.cpp: No such file or directory
g++.exe: fatal error: no input files
compilation terminated.
[Finished in 0.127s]
```
The `grammars.coffee` file c++ windows portion says:
```
args: (context) -> ["-c", "g++ -Wall -include stdio.h -include iostream '/mnt/" + path.posix.join.apply(path.posix, [].concat([context.filepath.split(path.win32.sep)[0].toLowerCase()], context.filepath.split(path.win32.sep).slice(1))).replace(":", "") + "' -o /tmp/cpp.out && /tmp/cpp.out"]
```
The target cpp file is located at d:\somefolder\some.cpp
Having no idea where the `C:/Program Files/Git` coming from?
And having no idea how to make the compilation work?
Any help appreciated!

**Pull request merged**